dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -537,14 +543,107 @@ private static void addBiases(Wmm memoryModel, EnumSet<Baseline> biases) {
-                .forEach(memoryModel::removeConstraint);
+    private static void removeFlaggedAxiomsIfNotNeeded(VerificationTask task) {
+        // We remove flagged axioms if we not check for them.
+        if (!task.getProperty().contains(Property.CAT_SPEC)) {
+            List.copyOf(task.getMemoryModel().getAxioms()).stream()
+                    .filter(Axiom::isFlagged)
+                    .forEach(task.getMemoryModel()::removeConstraint);
+        }
+    }
+
+    /*
+    The constraints/relations of the Wmm can be categorised into positive and negative,
+    depending on whether the number of negations applied to the constraint/relation is even (=positive)
+    or odd (=negative).
+    A relation can be both negative and positive if it is used multiple times,
+    once with odd negations and once with even negations.
+    It can also be neither, if the relation is dead (i.e., irrelevant for all axioms).
+     */
+    private record PolaritySeparator(Set<Constraint> positives, Set<Constraint> negatives) { }
+
+    private PolaritySeparator computePolaritySeparator(Wmm wmm) {
+        final Set<Constraint> positives = new HashSet<>();
+        final Set<Constraint> negatives = new HashSet<>();
+        final Constraint.Visitor<Void> collector = new Constraint.Visitor<>() {
+            private boolean polarity = true;
+
+            @Override
+            public Void visitAxiom(Axiom axiom) {
+                polarity = !axiom.isNegated();
+                process(axiom);
+                return axiom.getRelation().getDefinition().accept(this);
+            }
+
+            @Override
+            public Void visitDefinition(Definition def) {
+                if (process(def)) {
+                    def.getConstrainedRelations().subList(1, def.getConstrainedRelations().size())
+                            .forEach(r -> r.getDefinition().accept(this));
+                }
+                return null;
+            }
+
+            @Override
+            public Void visitDifference(Difference def) {
+                if (process(def)) {
+                    def.getMinuend().getDefinition().accept(this);
+                    polarity = !polarity;
+                    def.getSubtrahend().getDefinition().accept(this);
+                    polarity = !polarity;
+                }
+                return null;
+            }
+
+            private boolean process(Constraint c) {
+                return (polarity ? positives : negatives).add(c);
+            }
+        };
+
+        wmm.getAxioms().forEach(ax -> ax.accept(collector));
+        return new PolaritySeparator(positives, negatives);
+    }
+
+    /*
+        This pass rewrites the memory model so that no positive constraints appear on the rhs-side of a difference.
+        Suppose "r = a \ b" where r and a are negative and b is positive.
+        For example, this is the case if there is a constraint "~empty(r)"
+        We rewrite the model to
+            let b`= free()
```

Comment:
They relate in the sense that conceptually `b' == b`. It's just that we cannot enforce this equality eagerly because that would mean we need to encode `b` which is what we want to avoid. Instead, we learn the connection between those two variables lazily in CAAT (we actually do NOT learn the precise equality though).